### PR TITLE
Tracktable editing

### PR DESCRIPTION
**Pull request by daschuer**
This should improve the issues discussed here: 
https://github.com/mixxxdj/mixxx/pull/1550

By an extra checkbox "Edit metadata"

### CODE REVIEW

**Comment by Be-ing** [comment]
I don't understand how this helps. QAbstractItemView::SelectedClicked is mutually exclusive with double click actions.

**Comment by foss-** [comment]
The following link contains a video showing single and double click behavior on macOS 10.13.3. It shows 1x single click , 2x double click, 1x single click.
macOS waits for the second click before triggering the single click action.

For mixxx library behavior, my current preference would be to allow for a single click to trigger the edit track option (currently broken) and a double click to load the track into deck (working).

Editing track data is dangerous beacuse

1. it can happen by accident (and it did happen by accident, since mixxx was unable to differentiate between single and double click properly, resulting in double clicks triggering two options from what I understood). this for one is buggy behavior but it's also dangerous because edits can happen by accident very easily. since mixxx now allows writing metadata back to files, this is especially a serious problem - we don't want users to end up with a library full of tracks called "e" "gh" "c" etc) - see [LP1723920](https://bugs.launchpad.net/mixxx/+bug/1723920)
2. on macOS there is a bug where the waveform becomes unusable whenever edit track was used - see [LP1665583](https://bugs.launchpad.net/mixxx/+bug/1665583)

https://upload.disroot.org/r/xEpmOOVQ#iQUNBMkHhvHNJm0QpBpH56e+eAWDkeNYwbzyunySjwc=

Hope this helps finding a good solution for this little situation.

**Comment by Be-ing** [comment]
Thank you @foss-.

>  macOS waits for the second click before triggering the single click action.

I think this may be possible to implement this, but I think it would require not using QAbstractItemView::SelectedClicked.

**Comment by ywwg** [comment]
> I think this may be possible to implement this, but I think it would require not using QAbstractItemView::SelectedClicked. But I am doubtful we should rush such a hack.

I am really confused by your comments because the behavior @foss- describes is how mixxx was working before #1550.  I could doubleclick to load a track, or click twice to edit the metadata.  I never had a problem with double-click triggering metadata editing (in either linux or windows).  

My current suggestion is we revert #1550 and instead add a checkbox in the prefs to disable one-click track editing.

**Comment by ywwg** [comment]
@daschuer I get a build error on this code:

```
src/preferences/dialog/dlgpreflibrary.cpp: In member function ‘virtual void DlgPrefLibrary::slotUpdate()’:
src/preferences/dialog/dlgpreflibrary.cpp:178:13: error: ‘radioButton_dbclick_edit_metadata’ was not declared in this scope
             radioButton_dbclick_edit_metadata->setChecked(true);
             ^
src/preferences/dialog/dlgpreflibrary.cpp: In member function ‘virtual void DlgPrefLibrary::slotApply()’:
src/preferences/dialog/dlgpreflibrary.cpp:309:16: error: ‘radioButton_dbclick_edit_metadata’ was not declared in this scope
     } else if (radioButton_dbclick_edit_metadata->isChecked()) {
```

**Comment by daschuer** [comment]
> My current suggestion is we revert #1550 and instead add a checkbox in the prefs to disable one-click track editing.

That is what I am about to do. Sorry for the build error.

**Comment by ywwg** [comment]
For the record, if there is some bug where double-clicking is getting interpreted as click-on-selected and metadata editing is getting triggered, that's definitely a bug we should investigate.  What I'm confused by is I've never seen that before so I need more information about how the bug gets triggered (OS, double click speed settings, etc?)

**Comment by daschuer** [comment]
There is an issue only by design. The double click recognition works, but if you click too slow on a selected track you are in edit mode. This can most likely happen with a touch pad and the tab click is activated.

File: src/preferences/dialog/dlgpreflibrarydlg.ui [hunk 1]
```diff
@@ -249,6 +249,13 @@
         </property>
        </widget>
       </item>
+      <item row="4" column="0">
+       <widget class="QCheckBox" name="checkBoxEditMetadata">
+        <property name="text">
+         <string>Edit metadata</string>
```
Comment (Be-ing) [inline src/preferences/dialog/dlgpreflibrarydlg.ui]: This string should have more context. How about "Edit metadata after clicking selected track"?

File: src/preferences/dialog/dlgpreflibrary.cpp [hunk 2]
```diff
@@ -174,14 +171,15 @@ void DlgPrefLibrary::slotUpdate() {
...
-            radioButton_dbclick_edit_metadata->setChecked(true);
-            break;
     default:
             radioButton_dbclick_deck->setChecked(true);
             break;
     }
 
+    checkBoxEditMetadata->setChecked(
+            m_pConfig->getValue(
+                        ConfigKey("[Library]","EditMetadata"), true));
```
Comment (Be-ing) [inline src/preferences/dialog/dlgpreflibrary.cpp]: IMO this is much too dangerous to be enabled by default.

File: src/library/library.cpp [hunk 3]
```diff
@@ -172,6 +172,9 @@ Library::Library(
     } else {
         m_trackTableFont = QApplication::font();
     }
+
+    m_editMetadata = m_pConfig->getValue(
+            ConfigKey(kConfigGroup, "EditMetadata"), true);
```
Comment (Be-ing) [inline src/library/library.cpp]: The names in the code should have more context as well as the string shown to the user in the preferences. How about `m_editMetadataSelectedClick` for the boolean and `EditMetadataSelectedClick` for the ConfigKey name?

File: src/library/library.cpp [hunk 4]
```diff
@@ -412,16 +418,21 @@ QStringList Library::getDirs() {
...
     emit(setTrackTableFont(font));
 }
 
-void Library::slotSetTrackTableRowHeight(int rowHeight) {
+void Library::setRowHeight(int rowHeight) {
     m_iTrackTableRowHeight = rowHeight;
     emit(setTrackTableRowHeight(rowHeight));
 }
 
+void Library::setEditMedatata(bool enabled) {
```
Comment (Be-ing) [inline src/library/library.cpp]: setEditMetadataSelectedClick

File: src/preferences/dialog/dlgpreflibrary.cpp [hunk 5]
```diff
@@ -143,6 +139,7 @@ void DlgPrefLibrary::slotResetToDefaults() {
     checkBox_show_itunes->setChecked(true);
     checkBox_show_traktor->setChecked(true);
     radioButton_dbclick_bottom->setChecked(false);
+    checkBoxEditMetadata->setChecked(true);
```
Comment (Be-ing) [inline src/preferences/dialog/dlgpreflibrary.cpp]: checkBoxEditMetadataSelectedClicked

**Review by Be-ing — CHANGES_REQUESTED** [review]
This works as described. I still think it is a bad idea to have both a SelectedClicked action and a double click action. But as long as I have the option to turn off the SelectedClick action, that's good enough for me.

File: src/preferences/dialog/dlgpreflibrarydlg.ui (hunk 1 above)
Comment (daschuer) [reply src/preferences/dialog/dlgpreflibrarydlg.ui]: For 2.1, I would like to keep this text, because we already have translations for it. 
It is not too bad and details can go to a Tooltip 
```
This enables instant metadata editing within the track table. Editing starts when clicking on an already selected item.
```

File: src/preferences/dialog/dlgpreflibrary.cpp (hunk 2 above)
Comment (daschuer) [reply src/preferences/dialog/dlgpreflibrary.cpp]: Since this was the default in 2.0 I like to keep it. Windows and Mac user are used to that "feature"

File: src/preferences/dialog/dlgpreflibrarydlg.ui (hunk 1 above)
Comment (esbrandt) [reply src/preferences/dialog/dlgpreflibrarydlg.ui]: The only recent added string is translated just for ``DE``so far. You can easily change it again.

Comment (daschuer) [reply src/preferences/dialog/dlgpreflibrarydlg.ui]: OK thank you.

**Comment by daschuer** [comment]
Done.

File: src/preferences/dialog/dlgpreflibrarydlg.ui [hunk 6]
```diff
@@ -249,14 +249,14 @@
         </property>
        </widget>
       </item>
-      <item row="4" column="0">
-       <widget class="QCheckBox" name="checkBoxEditMetadata">
+      <item row="4" column="0" colspan="2">
+       <widget class="QCheckBox" name="checkBoxEditMetadataSelectedClicked">
         <property name="toolTip">
          <string>This enables instant metadata editing within the track table.
 Editing starts when clicking on an already selected item.</string>
```
Comment (Be-ing) [inline src/preferences/dialog/dlgpreflibrarydlg.ui]: I think this tooltip is unnecessary with the updated string.

**Comment by daschuer** [comment]
Ready

File: src/preferences/dialog/dlgpreflibrary.cpp (hunk 2 above)
Comment (Be-ing) [reply src/preferences/dialog/dlgpreflibrary.cpp]: I don't think either of those are good reasons to make this on by default. This behavior is so unexpected and confusing that no one could even describe quite what the problem was until I started working on the code and reading the Qt documentation. If users have to read the "Edit metadata after clicking selected track" description in the preferences before enabling this behavior, I think they will have an easier time understanding how it works and therefore understand how to use it intentionally and avoid activating it unintentionally.

**Comment by Be-ing** [comment]
LGTM except for the option being enabled by default.

File: src/preferences/dialog/dlgpreflibrary.cpp (hunk 2 above)
Comment (daschuer) [reply src/preferences/dialog/dlgpreflibrary.cpp]: For feeling is that this topic has stirred up, because of a single users bug report. 
All Windows an Mac users expect the feature and can handle it in their file manager. 
This feature was enabled for ages in Mixxx, without anyone else complaining.
But we had complains during the short phase with the double click edit, loosing this feature. 

So I assume that many user would feel the disabled feature as a regression. It might took a while to discover the check-box to re-enable it.

I do not use this feature and I also have never accidentally enabled it. So I do not care personally about the default. 

How do others think?

Comment (foss-) [reply src/preferences/dialog/dlgpreflibrary.cpp]: @daschuer are you referring to https://bugs.launchpad.net/mixxx/+bug/1665583 ? If so, let me reassure you, this has been giving mac users a hard time using mixxx to the point where live usage becomes questionable due to unusable waveforms when accidentally getting into track edit mode. The bug was files feb 2017 and only disabling edit mode / adding an option in prefs makes plex usable for macOS users on a laptop with trackpad. Let me know if I am missing something. My experience is limited to macOS, so this may be working great on linux / win.

Comment (ronso0) [reply src/preferences/dialog/dlgpreflibrary.cpp]: For me this never was an issue, maybe because I'm on Linux and because I don't have to use keyboard shortcuts. Direct editing is a great feature, I use it a lot for writing track comments.

Let's say I run into the issue that track editing interferes with my track load workflow and annoys me, I'd definitely dig through the Preferences options and see if I can disable it. This is definitely biased since I know how much Mixxx can be customized.

Or put the other way around (off by default): I wouldn't expect that feature and probably wouldn't miss it, so I wouldn't go looking for it.

So I vote for default=ON

Comment (Be-ing) [reply src/preferences/dialog/dlgpreflibrary.cpp]: >  All Windows an Mac users expect the feature and can handle it in their file manager.

I've never used Mac very much, but long ago when I used to use Windows I do not remember this click selected item to edit behavior. I remember pressing F2 to rename files. Also, the use cases of file managers and a DJ program are quite different. File managers are not used for live performance.

Comment (daschuer) [reply src/preferences/dialog/dlgpreflibrary.cpp]: I was reffering https://bugs.launchpad.net/mixxx/+bug/1743238

Comment (Be-ing) [reply src/preferences/dialog/dlgpreflibrary.cpp]: Note that https://bugs.launchpad.net/mixxx/+bug/1743238 has 4 people marked as being affected. It is not just one user opening a bug report.

Comment (foss-) [reply src/preferences/dialog/dlgpreflibrary.cpp]: @daschuer In that case please have a look at the bug linked under "2." which elaborates on a nasty side effects of edit mode on macOS. I do not want to over-complicate this discussion. Having an option to disable editing is just fine for me, thus I do not have strong feelings about the default.

I currently see two concerns regarding defaulting to on:
1. writing to files may lead to unwanted results, i.e. having files named "e" "ff" or whatever user types when being in edit mode accidentally. I can confirm that 100% of the time I went to edit mode it was by accident. (could be solved by having write to files disabled by default and show an informative warning when first enabled - both is the case currently iirc?)
2. macOS side effect https://bugs.launchpad.net/mixxx/+bug/1665583

Comment (Be-ing) [reply src/preferences/dialog/dlgpreflibrary.cpp]: As if on cue, we have [another macOS user reporting the same bug](https://bugs.launchpad.net/mixxx/+bug/1759433).

**Comment by ywwg** [comment]
I'll check this tonight

File: src/preferences/dialog/dlgpreflibrarydlg.ui [hunk 7]
```diff
@@ -249,6 +249,13 @@
         </property>
        </widget>
       </item>
+      <item row="4" column="0" colspan="2">
+       <widget class="QCheckBox" name="checkBoxEditMetadataSelectedClicked">
+        <property name="text">
+         <string>Edit metadata after clicking selected track</string>
```
Comment (ywwg) [inline src/preferences/dialog/dlgpreflibrarydlg.ui]: How about:  "Allow editing of metadata by clicking selected track"

I wonder if this is a good candidate for an inverted bool, because the major use-case is "locking" the library so it can't accidentally be edited: "Prevent editing of metadata by clicking selected track"

File: src/library/dlganalysis.cpp [hunk 8]
```diff
@@ -62,6 +62,13 @@ DlgAnalysis::DlgAnalysis(QWidget* parent,
             SIGNAL(selectionChanged(const QItemSelection &, const QItemSelection&)),
             this,
             SLOT(tableSelectionChanged(const QItemSelection &, const QItemSelection&)));
+
+    connect(pLibrary, SIGNAL(setTrackTableFont(QFont)),
```
Comment (ywwg) [inline src/library/dlganalysis.cpp]: is this an unrelated change?  I still don't see the correct font being used in the analyze view of the library.  (That's been broken for a long time)

**Comment by ywwg** [comment]
The changes works as described and I like this solution a lot, thanks!  I just want to make sure the preference text is right before giving LGTM

File: src/preferences/dialog/dlgpreflibrary.cpp (hunk 2 above)
Comment (ywwg) [reply src/preferences/dialog/dlgpreflibrary.cpp]: I'm confused, the bug was that "it's too easy to trigger metadata editing" but this user is reporting "If I click any field in the library (say, to change a song's name/artist), the UI goes down to 2 FPS."  How is that the same bug?

File: src/preferences/dialog/dlgpreflibrarydlg.ui (hunk 7 above)
Comment (daschuer) [reply src/preferences/dialog/dlgpreflibrarydlg.ui]: Yes, good idea.

File: src/preferences/dialog/dlgpreflibrary.cpp (hunk 2 above)
Comment (ywwg) [reply src/preferences/dialog/dlgpreflibrary.cpp]: > macOS side effect

This sounds like a separate issue that needs to be addressed on that platform.  Disabling click-to-edit would just be sweeping the problem under the carpet.

Comment (ywwg) [reply src/preferences/dialog/dlgpreflibrary.cpp]: > How do others think?

We've had click-to-edit for many years without complaints.  I think the problem is important enough to create the preference that we have done here, but I feel comfortable defaulting to the existing behavior.  We can update the bug and point people to the new preference.

It's very hard to get a sense of what percentage of users a particular mis-feature affects, and mixxx has a habit of implementing features for single-digit user counts vs the ~million downloads we get per year.  Sure, not everyone who experiences a problem will register for a bug account and find the same bug, so at some point we need to use our own judgement.

**Comment by daschuer** [comment]
OK, checkbox inverted and analysis feature connections repaired.

File: src/preferences/dialog/dlgpreflibrary.cpp (hunk 2 above)
Comment (Be-ing) [reply src/preferences/dialog/dlgpreflibrary.cpp]: > We've had click-to-edit for many years without complaints.

It is a little frustrating to have to repeat myself:
> Note that https://bugs.launchpad.net/mixxx/+bug/1743238 has 4 people marked as being affected. It is not just one user opening a bug report.

Comment (Be-ing) [reply src/preferences/dialog/dlgpreflibrary.cpp]: > This sounds like a separate issue that needs to be addressed on that platform. Disabling click-to-edit would just be sweeping the problem under the carpet.

Well we have no macOS developers and we are try to get a release out imminently. This is the time for quick hacks, not debugging complicated issues.

Comment (ywwg) [reply src/preferences/dialog/dlgpreflibrary.cpp]: I'm aware that the bug exists, but it was reported in January of this year and we've had this feature for many years, so this seems to be a new problem on OSX only.  I suspect something changed on that platform, like the addition of "force clicks" which is VERY confusing for users who don't know what it is (including myself).  Maybe "force click" is enabling edit mode?

What about making the default off for OSX only?

Comment (Be-ing) [reply src/preferences/dialog/dlgpreflibrary.cpp]: The UI lagging with editing metadata from the track table is only on macOS, but the high risk of accidentally editing metadata is for all OSs. This has been a problem for me on GNU/Linux with Mixxx for as long as I've been using it. I didn't report a bug for it earlier because the behavior is so unintuitive that I couldn't figure out exactly what triggered the accidental edits.

Having different defaults for different platforms doesn't seem like a great idea to me.

Comment (ywwg) [reply src/preferences/dialog/dlgpreflibrary.cpp]: Fair enough.  My preference is to leave it on by default but as long as there's adequate documentation I'm ok with the default being either way

Comment (Be-ing) [reply src/preferences/dialog/dlgpreflibrary.cpp]: If the default is changed to off, I think the new inverted "Prevent editing of metadata by clicking selected track" wording is odd.

**Comment by ywwg** [comment]
> LGTM except for the option being enabled by default.

It is my dream that Mixxx can learn to decide issues of opinion through a consensus process instead of a battle of wills.  It's very stressful when some people are saying "this is what I prefer let's find a solution we can all live with" and other people are saying "I refuse to LGTM unless it's done this way."

**Comment by daschuer** [comment]
I did not meant to treat you disrespectful, I am sorry for this. We already had an agreement that this PR is ready for merge, except the default question. The default can be changed at any time, no reason to delay this and have an untranslated string for this feature. 

This can be treated as a workaround for https://bugs.launchpad.net/mixxx/+bug/1665583, with any default value. But this bug discovers an other strong issue in our Mixxx project in general:
If we are not able to investigate this party stopping bug since 2017-02-17, how could we do this now in reasonable time before the scheduled release. We have simply no one, who is able to reproduce it and fix this issue. Is is questionable if we are able to support Mac OS any longer. 
Maybe we should drop the official Mac Support until we find an engaged contributor.

**Comment by Be-ing** [comment]
"LGTM except" does not mean "LGTM". The default is important, especially considering the consequences of this.

> Maybe we should drop the official Mac Support until we find an engaged contributor.

I was thinking about that too, but I think that is the opposite of what we should do. I think if we do that, then there's a decent chance we'll never have a Mac contributor again. We have had several Mac developers express interest in contributing lately but all of them have been discouraged by out-of-date and incomplete documentation. I will try to borrow a Mac tomorrow to look into the macOS issues.

**Comment by daschuer** [comment]
Thank you very much. I hope that we see clear after your investigations.

**Comment by foss-** [comment]
Dropping macOS support will send wrong signals and minimize the chance for macOS devs to join in. I've been using mixxx and closely observing it's development for a while now. It's made amazing progress. There are a few serious macOS related bugs. But regardless of the default for this, adding this option will make mixxx a lot more usable on macOS.

So while there are some intense debates going on currently the small mixxx dev team has a lot on their plate to handle and I think the team is doing an amazing job.

Thank you all for your work and dedicating your time to this amazing software <3

**Comment by poelzi** [comment]
To give my 2cents. Whenever I have the edit functionality open, it's by accident. Mixxx is not a good tag management software and does not have to be in my opinion. It's nice if you can fix som mistakes you find while using it, but having such a easy accessible edit functionality is more annoying then it helps.
I'm very happy that I can deactivate it soon.
